File: src/main/java/me/zeroeightsix/kami/module/modules/misc/AntiAFK.java
package me.zeroeightsix.kami.module.modules.misc;

import baritone.api.BaritoneAPI;
import baritone.api.pathing.goals.GoalXZ;
import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.Timer;
import net.minecraft.network.play.client.CPacketAnimation;
import net.minecraft.network.play.server.SPacketChat;
import net.minecraft.util.EnumHand;
import net.minecraftforge.fml.common.gameevent.InputEvent;

import java.util.Objects;
import java.util.Random;

import static me.zeroeightsix.kami.util.MathsUtils.reverseNumber;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendServerMessage;

/**
 * Created by 086 on 16/12/2017.
 * Updated by dominikaaaa on 21/04/20
 * Updated by Xiaro on 23/08/20
 * TODO: Path finding to stay inside 1 chunk
 * TODO: Render which chunk is selected
 */
@Module.Info(
        name = "AntiAFK",
        category = Module.Category.MISC,
        description = "Prevents being kicked for AFK"
)
public class AntiAFK extends Module {

    private final Setting<Integer> frequency = register(Settings.integerBuilder("ActionFrequency").withMinimum(1).withMaximum(100).withValue(40).build());
    public Setting<Boolean> autoReply = register(Settings.b("AutoReply", true));
    private final Setting<Boolean> swing = register(Settings.b("Swing", true));
    private final Setting<Boolean> jump = register(Settings.b("Jump", true));
    private final Setting<Boolean> squareWalk = register(Settings.b("SquareWalk", true));
    private final Setting<Integer> radius = register(Settings.integerBuilder("Radius").withMinimum(1).withValue(64).withVisibility(v -> squareWalk.getValue()).build());
    private final Setting<Boolean> turn = register(Settings.b("Turn", true));
    private final Setting<Integer> inputTimeout = register(Settings.integerBuilder("InputTimeout(m)").withValue(0).withRange(0, 15).build());

    private final Random random = new Random();

    private final int[] squareStartCoords = {0, 0};
    private int squareStep = 0;
    private Boolean baritoneDisconnectOnArrival = false;
    private final Timer inputTimer = new Timer(Timer.TimeUnit.MINUTES);

    public AntiAFK() {
        super();

        squareWalk.settingListener = setting -> {
            if (isEnabled()) {
                BaritoneAPI.getProvider().getPrimaryBaritone().getPathingBehavior().cancelEverything();
            }
        };
    }

    @Override
    public void onEnable() {
        if (mc.player == null || inputTimeout.getValue() != 0)
            return;

        baritoneDisconnectOnArrival = BaritoneAPI.getSettings().disconnectOnArrival.value;
        BaritoneAPI.getSettings().disconnectOnArrival.value = false;
        squareStartCoords[0] = (int) mc.player.posX;
        squareStartCoords[1] = (int) mc.player.posZ;
    }

    @Override
    public void onDisable() {
        if (mc.player == null)
            return;

        BaritoneAPI.getSettings().disconnectOnArrival.value = baritoneDisconnectOnArrival;
        if (isBaritoneActive() && squareWalk.getValue())
            BaritoneAPI.getProvider().getPrimaryBaritone().getPathingBehavior().cancelEverything();
    }

    @Override
    public void onUpdate() {
        if (inputTimeout.getValue() != 0) {
            if (!squareWalk.getValue() && BaritoneAPI.getProvider().getPrimaryBaritone().getPathingControlManager().mostRecentInControl().isPresent()) {
                inputTimer.reset();
            }
            if (!inputTimer.tick(inputTimeout.getValue(), false)) {
                return;
            }
        }

        if (mc.playerController.getIsHittingBlock()) return;

        if (swing.getValue() && mc.player.ticksExisted % (0.5 * getFrequency()) == 0) {
            Objects.requireNonNull(mc.getConnection()).sendPacket(new CPacketAnimation(EnumHand.MAIN_HAND));
        }

        if (squareWalk.getValue() && mc.player.ticksExisted % getFrequency() == 0 && !isBaritoneActive()) {
            int r = clamp(radius.getValue());
            switch (squareStep) {
                // +z
                case 0:
                    baritoneGotoXZ(squareStartCoords[0], squareStartCoords[1] + r);
                    break;
                // +x
                case 1:
                    baritoneGotoXZ(squareStartCoords[0] + r, squareStartCoords[1] + r);
                    break;
                // -z
                case 2:
                    baritoneGotoXZ(squareStartCoords[0] + r, squareStartCoords[1]);
                    break;
                // -x
                case 3:
                    baritoneGotoXZ(squareStartCoords[0], squareStartCoords[1]);
                    break;
            }
            squareStep = (squareStep + 1) % 4;
        }

        if (jump.getValue() && mc.player.ticksExisted % (2 * getFrequency()) == 0) {
            mc.player.jump();
        }

        if (turn.getValue() && mc.player.ticksExisted % (0.375 * getFrequency()) == 0) {
            mc.player.rotationYaw = random.nextInt(360) - makeNegRandom(180);
        }
    }

    @Override
    public String getHudInfo() {
        return String.valueOf(System.currentTimeMillis() - inputTimer.lastTickTime);
    }

    @EventHandler
    private final Listener<PacketEvent.Receive> receiveListener = new Listener<>(event -> {
        if (autoReply.getValue() && event.getPacket() instanceof SPacketChat && ((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().contains("whispers: ") && !((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().contains(mc.player.getName())) {
            sendServerMessage("/r I am currently AFK and using KAMI Blue!");
        }
    });

    @EventHandler
    private final Listener<InputEvent.MouseInputEvent> mouseInputListener = new Listener<>(event -> {
        if (inputTimeout.getValue() != 0 && isInputting()) {
            inputTimer.reset();
        }
    });

    @EventHandler
    private final Listener<InputEvent.KeyInputEvent> keyInputListener = new Listener<>(event -> {
        if (inputTimeout.getValue() != 0 && isInputting()) {
            inputTimer.reset();
        }
    });

    private boolean isInputting() {
        return mc.gameSettings.keyBindAttack.isKeyDown()
                || mc.gameSettings.keyBindUseItem.isKeyDown()
                || mc.gameSettings.keyBindJump.isKeyDown()
                || mc.gameSettings.keyBindSneak.isKeyDown()
                || mc.gameSettings.keyBindForward.isKeyDown()
                || mc.gameSettings.keyBindBack.isKeyDown()
                || mc.gameSettings.keyBindLeft.isKeyDown()
                || mc.gameSettings.keyBindRight.isKeyDown();
    }

    private boolean isBaritoneActive() {
        return BaritoneAPI.getProvider().getPrimaryBaritone().getCustomGoalProcess().isActive();
    }

    private void baritoneGotoXZ(int x, int z) {
        BaritoneAPI.getProvider().getPrimaryBaritone().getCustomGoalProcess().setGoalAndPath(new GoalXZ(x, z));
    }

    private int clamp(int val) {
        return Math.max(val, 0);
    }

    private float getFrequency() {
        return reverseNumber(frequency.getValue(), 1, 100);
    }

    private int makeNegRandom(int input) {
        int rand = random.nextBoolean() ? 1 : 0;
        if (rand == 0) {
            return -input;
        } else {
            return input;
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/misc/AntiAFK.kt
package me.zeroeightsix.kami.module.modules.misc

import baritone.api.pathing.goals.GoalXZ
import me.zeroeightsix.kami.event.SafeClientEvent
import me.zeroeightsix.kami.event.events.BaritoneSettingsInitEvent
import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.module.Category
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.settings.impl.primitive.BooleanSetting
import me.zeroeightsix.kami.util.BaritoneUtils
import me.zeroeightsix.kami.util.TickTimer
import me.zeroeightsix.kami.util.TimeUnit
import me.zeroeightsix.kami.util.text.MessageDetection
import me.zeroeightsix.kami.util.text.MessageSendHelper.sendServerMessage
import me.zeroeightsix.kami.util.threads.safeListener
import net.minecraft.network.play.server.SPacketChat
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.gameevent.InputEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import org.kamiblue.event.listener.listener
import kotlin.random.Random

/**
 * TODO: Path finding to stay inside 1 chunk
 * TODO: Render which chunk is selected
 */
internal object AntiAFK : Module(
    name = "AntiAFK",
    category = Category.MISC,
    description = "Prevents being kicked for AFK"
) {
    private val delay by setting("ActionDelay", 50, 5..100, 5)
    private val variation by setting("Variation", 25, 0..50, 5)
    private val autoReply by setting("AutoReply", true)
    private val swing = setting("Swing", true)
    private val jump = setting("Jump", true)
    private val turn = setting("Turn", true)
    private val walk = setting("Walk", true)
    private val radius by setting("Radius", 64, 8..128, 8)
    private val inputTimeout by setting("InputTimeout(m)", 0, 0..15, 1)
    private val allowBreak by setting("AllowBreakingBlocks", false, { walk.value })

    private var startPos: BlockPos? = null
    private var squareStep = 0
    private var baritoneAllowBreak = false
    private var baritoneDisconnectOnArrival = false
    private val inputTimer = TickTimer(TimeUnit.MINUTES)
    private val actionTimer = TickTimer(TimeUnit.TICKS)
    private var nextActionDelay = 0

    override fun getHudInfo(): String {
        return if (inputTimeout == 0) ""
        else ((System.currentTimeMillis() - inputTimer.time) / 1000L).toString()
    }

    init {
        onEnable {
            baritoneAllowBreak = BaritoneUtils.settings?.allowBreak?.value ?: true
            if (!allowBreak) BaritoneUtils.settings?.allowBreak?.value = false
            inputTimer.reset()
            baritoneDisconnectOnArrival()
        }

        onDisable {
            startPos = null
            BaritoneUtils.settings?.allowBreak?.value = baritoneAllowBreak
            BaritoneUtils.settings?.disconnectOnArrival?.value = baritoneDisconnectOnArrival
            BaritoneUtils.cancelEverything()
        }

        listener<BaritoneSettingsInitEvent> {
            baritoneDisconnectOnArrival()
        }
    }

    private fun baritoneDisconnectOnArrival() {
        BaritoneUtils.settings?.disconnectOnArrival?.let {
            baritoneDisconnectOnArrival = it.value
            it.value = false
        }
    }

    init {
        listener<PacketEvent.Receive> {
            if (!autoReply || it.packet !is SPacketChat) return@listener
            if (MessageDetection.Direct.RECEIVE detect it.packet.chatComponent.unformattedText) {
                sendServerMessage("/r I am currently AFK and using KAMI Blue!")
            }
        }

        listener<InputEvent.MouseInputEvent> {
            if (inputTimeout != 0 && isInputting()) {
                inputTimer.reset()
            }
        }

        listener<InputEvent.KeyInputEvent> {
            if (inputTimeout != 0 && isInputting()) {
                inputTimer.reset()
            }
        }
    }

    private fun isInputting() =
        mc.gameSettings.keyBindAttack.isKeyDown
            || mc.gameSettings.keyBindUseItem.isKeyDown
            || mc.gameSettings.keyBindJump.isKeyDown
            || mc.gameSettings.keyBindSneak.isKeyDown
            || mc.gameSettings.keyBindForward.isKeyDown
            || mc.gameSettings.keyBindBack.isKeyDown
            || mc.gameSettings.keyBindLeft.isKeyDown
            || mc.gameSettings.keyBindRight.isKeyDown

    init {
        safeListener<TickEvent.ClientTickEvent> {
            if (inputTimeout != 0) {
                if (BaritoneUtils.isActive) {
                    inputTimer.reset()
                } else if (!inputTimer.tick(inputTimeout.toLong(), false)) {
                    startPos = null
                    return@safeListener
                }
            }

            if (actionTimer.tick(nextActionDelay.toLong())) {
                val random = if (variation > 0) (0..variation).random() else 0
                nextActionDelay = delay + random

                when ((getAction())) {
                    Action.SWING -> player.swingArm(EnumHand.MAIN_HAND)
                    Action.JUMP -> player.jump()
                    Action.TURN -> player.rotationYaw = Random.nextDouble(-180.0, 180.0).toFloat()
                }

                if (walk.value && !BaritoneUtils.isActive) {
                    squareWalk()
                }
            }
        }
    }

    private fun getAction(): Action? {
        if (!swing.value && !jump.value && !turn.value) return null
        val action = Action.values().random()
        return if (action.setting.value) action else getAction()
    }

    private fun SafeClientEvent.squareWalk() {
        if (startPos == null) startPos = player.position

        startPos?.let {
            when (squareStep) {
                0 -> baritoneGotoXZ(it.x, it.z + radius)
                1 -> baritoneGotoXZ(it.x + radius, it.z + radius)
                2 -> baritoneGotoXZ(it.x + radius, it.z)
                3 -> baritoneGotoXZ(it.x, it.z)
            }
        }

        squareStep = (squareStep + 1) % 4
    }

    private fun baritoneGotoXZ(x: Int, z: Int) {
        BaritoneUtils.primary?.customGoalProcess?.setGoalAndPath(GoalXZ(x, z))
    }

    private enum class Action(val setting: BooleanSetting) {
        SWING(swing),
        JUMP(jump),
        TURN(turn)
    }

    init {
        walk.listeners.add {
            BaritoneUtils.cancelEverything()
        }
    }
}